libqrexec/ioall.c

```diff
@@ -90,6 +98,66 @@ int read_all(int fd, void *buf, int size)
     return 1;
 }
 
+// FIXME: this code is seemingly-correct but needs careful review
+void *qubes_read_all_to_malloc(int fd, size_t initial_buffer_size, size_t max_bytes, size_t *len)
+{
+    size_t buf_size = initial_buffer_size;
+    size_t offset = 0;
+    if (buf_size < 2 || buf_size > (size_t)PTRDIFF_MAX) {
+        LOG(ERROR,
+            "Minimum buf size must between 2 and %zd inclusive, got %zu",
+            PTRDIFF_MAX,
+            buf_size);
+        abort();
+    }
+    char *buf = malloc(buf_size);
+    if (buf == NULL) {
+        LOG(ERROR, "malloc() for %zu bytes failed!", buf_size);
+        abort();
+    }
+    *len = 0;
+    for (;;) {
+        size_t to_read = buf_size - 1 - offset;
+        ssize_t res = read(fd, buf + offset, to_read > INT_MAX ? INT_MAX : to_read);
```

Comment:
I'd prefer to cap `max_bytes` on INT_MAX (this function is ever called with max_bytes at most 64k), than trying to handle such corner case.

Reply:
Done, along with adding lots of comments to this function.